packages/SwingSet/src/kernel/dynamicVat.js

```diff
@@ -13,34 +14,56 @@ export function makeDynamicVatCreator(stuff) {
    * creates the vat and sends a notification to the device. The root object
    * will be available soon, but we immediately return the vatID so the ultimate
    * requestor doesn't have to wait.
    *
-   * @param vatSourceBundle a source bundle (JSON-serializable data) which
-   * defines the vat. This should be generated by calling bundle-source on a
-   * module whose default export is makeRootObject(), which takes E as a
-   * parameter and returns a root object.
-   * @param options an options bundle. The only option defined so far is
-   * 'metered', which defaults to 'true'. If 'true', the new dynamic vat is
-   * subject to a meter that limits the amount of computation and allocation
-   * that can occur during any given crank. Stack frames are limited as well.
-   * The meter is refilled between cranks, but if the meter ever underflows,
-   * the vat is terminated. If 'false', the vat is unmetered.
+   * @param source an object which either has a `bundle` (JSON-serializable
+   * data) or a `bundleName` string. The bundle defines the vat, and should
+   * be generated by calling bundle-source on a module with an export named
+   * makeRootObject(). If `bundleName` is used, it must identify a bundle
+   * already known to the kernel (via the `config.bundles` table).
+   * @param options a "dynamicOptions" bundle. This recognizes two options so
+   * far. The first is 'metered' (defaults to 'true') which subjects the new
+   * dynamic vat to a meter that limits the amount of computation and
+   * allocation that can occur during any given crank. Stack frames are
+   * limited as well. The meter is refilled between cranks, but if the meter
+   * ever underflows, the vat is terminated. If 'false', the vat is
+   * unmetered. The other is 'vatParameters', which provides the contents of
+   * the second argument to 'buildRootObject()'
    *
    * @return { vatID } the vatID for a newly created vat. The success or
    * failure of the operation will be reported in a message to the admin vat,
    * citing this vatID
    */
 
-  function createVatDynamically(vatSourceBundle, dynamicOptions = {}) {
+  function createVatDynamically(source, dynamicOptions = {}) {
+    const vatID = allocateUnusedVatID();
+    kernelKeeper.addDynamicVatID(vatID);
+    const vatKeeper = kernelKeeper.allocateVatKeeperIfNeeded(vatID);
+    vatKeeper.setSourceAndOptions(source, dynamicOptions);
+    // eslint-disable-next-line no-use-before-define
+    return create(vatID, source, dynamicOptions, true);
+  }
+
+  function recreateVatDynamically(vatID, source, dynamicOptions) {
+    // eslint-disable-next-line no-use-before-define
+    return create(vatID, source, dynamicOptions, false);
+  }
+
+  function create(vatID, source, dynamicOptions, notifyNewVat) {
+    const vatSourceBundle = source.bundle || getBundle(source.bundleName);
+    if (!vatSourceBundle) {
+      throw Error(`Bundle ${source.bundleName} not found`);
```

Comment:
This will be ugly if `source` has neither `bundle` nor `bundleName`.  Should either validate the `source` parameter or switch back to the 2-method API.